lib/Doctrine/ORM/Tools/SchemaTool.php

```diff
@@ -274,6 +275,15 @@ public function getSchemaFromMetadata(array $classes)
 
             if (isset($class->table['uniqueConstraints'])) {
                 foreach ($class->table['uniqueConstraints'] as $indexName => $indexData) {
+                    $uniqIndex = new Index($indexName, $indexData['columns'], true, false, [], isset($indexData['options']) ? $indexData['options'] : []);
+
+                    foreach ($table->getIndexes() as $tableIndexName => $tableIndex) {
+                        if ($tableIndex->isFullfilledBy($uniqIndex)) {
+                            $table->dropIndex($tableIndexName);
+                            break;
```

Comment:
IMO the preference should be `uniqueConstraint` (explicit) > `unique=true` (implicit) if and only if `uniqueConstraint` spans ONLY the `unique=true` column (same behaviour as for `uniqueConstraint` (explicit) > `foreignKey` index (implicit).
Not sure about OneToOne but it should be similar I guess.

So what about simply leaving this block as-is and move it to the top of the classes loop to line 162. Then it will first add explicit unique constraints before adding implicit unique indexes from the columns gathering. `gatherColumn()` should then ask the table is a fullfilling unique index is already available and skip addition eventually.